Bluetooth: Controller: Fake random number generation in HAL

cvinayak opened the pull request:
Add a very simple fake random number generation in HAL with possible entropy based on access address reception timing.

This PR provides a fake random number generation to avoid returning uninitialized memory which is most cases returns the same uninitialized values (of course, isnt random any more).

```
/* FIXME: No suitable entropy device available yet.
 *        It is required by Controller to use random numbers.
 *        Hence, return uninitialized buf contents, for now.
*/
```

cvinayak committed:
Bluetooth: Controller: Fake random number generation in HAL

Add a very simple fake random number generation in HAL with
possible entropy based on access address reception timing.

This change provides a fake random number generation to
avoid returning uninitialized memory, which in most cases
returns the same uninitialized values (of course, isn't
random any more).

There are code comments like below, that needs to be
correctly addressed in the future:
```
/* FIXME: No suitable entropy device available yet.
 *        It is required by Controller to use random numbers.
 *        Hence, return uninitialized buf contents, for now.
*/
```

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

ajf58 commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -297,6 +297,16 @@ config BT_CTLR_ADVANCED_FEATURES
+	  A fake random number generator provided by the Controller to meet the
+	  requirement of advertising random delay and cryptography needs. This
+	  option makes the Controller experimental. This random number generator
+	  may not meet the FIPS PUB 140-2 compliance.
```
```suggestion
	  definitely does not meet the FIPS PUB 140-2 compliance.
```

cvinayak replied:
Sure :)

ajf58 commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1179,6 +1179,50 @@ uint32_t radio_bc_has_match(void)
+{
+	uint8_t ms;
+
+	/* This is something that came of the guy writing it and who saw that
+	 * the sequence of returned value in the broadcaster sample seemed
+	 * random enough for him!
+	 */
+	ms = rng_fake >> 4;
+	rng_fake <<= 1;
+	rng_fake ^= (uint8_t)(value & 0xFF);
+	rng_fake ^= (uint8_t)((value >> 8) & 0xFF);
+	rng_fake ^= (uint8_t)((value >> 16) & 0xFF);
+	rng_fake ^= (uint8_t)((value >> 24) & 0xFF);
+	rng_fake ^= ~ms;
+
+	return rng_fake;
```
There are a lot of well-documented (and analyzed) pseudo-random number generators. This look a bit like a LFSR. Can you help provide a bit more explanation of this generator?

cvinayak replied:
Explained in the comment already in code.

jhedberg commented:
It'd likely be best to accompany this with something like what's used for similar features elsewhere:
https://github.com/zephyrproject-rtos/zephyr/blob/50b07f9480a49f2b489ade9dae98b91fc3f2a643/subsys/random/CMakeLists.txt#L11-L16

cvinayak commented:
> It'd likely be best to accompany this with something like what's used for similar features elsewhere:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/50b07f9480a49f2b489ade9dae98b91fc3f2a643/subsys/random/CMakeLists.txt#L11-L16

Agree. The implementation in the PR isvery crude, added for testing purpose due to current none availability (until a crypto secure random number generations is made available).

cvinayak commented:
> It'd likely be best to accompany this with something like what's used for similar features elsewhere:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/50b07f9480a49f2b489ade9dae98b91fc3f2a643/subsys/random/CMakeLists.txt#L11-L16

Hmmm... on second thought why is this timer generator not available for nRF54L.... thinking...

Thalley commented:
The commit message does not state why this is being added at all, and there are no feature request nor bug report to lean on. 

Why is this being added? What are the requirements/use cases?

cvinayak commented:
> The commit message does not state why this is being added at all, and there are no feature request nor bug report to lean on.
> 
> Why is this being added? What are the requirements/use cases?

Updated the description... 

> It'd likely be best to accompany this with something like what's used for similar features elsewhere:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/50b07f9480a49f2b489ade9dae98b91fc3f2a643/subsys/random/CMakeLists.txt#L11-L16

Done.



> The commit message does not state why this is being added at all, and there are no feature request nor bug report to lean on.
> 
> Why is this being added? What are the requirements/use cases?

Updated commit message.

Thalley commented on subsys/bluetooth/controller/CMakeLists.txt:
```diff
@@ -1,11 +1,17 @@
                   "Controller's functional behavior.")
 endif()
 
+if (CONFIG_BT_CTLR_RNG_FAKE)
+  message(WARNING "\nCONFIG_BT_CTLR_RNG_FAKE is not a truly random generator. "
+                  "This capability is not secure and it is provided "
+                  "for testing purposes only. Use it carefully.")
```
Suggest to also add `CONFIG_BT_CTLR_RNG_FAKE` to the `hardened.csv` file

cvinayak replied:
How come `EXPERIMENTAL` is not part of `hardened.csv`?

Thalley replied:
No idea :)

Thalley reviewed (commented):
The commit/PR still does not explain _why_ this is being added, and why a random fake is being added and why we cannot use a more proper random number generator.

cvinayak commented:
> The commit/PR still does not explain _why_ this is being added, and why a random fake is being added and why we cannot use a more proper random number generator.

`No suitable entropy device available yet.`  I do not know the plan of an open source cryptographic secure random number generation support, either.

Thalley commented:
> > The commit/PR still does not explain _why_ this is being added, and why a random fake is being added and why we cannot use a more proper random number generator.
> 
> `No suitable entropy device available yet.` I do not know the plan of an open source cryptographic secure random number generation support, either.

Didn't consider an example reference to a in-line code comment as a reason :D 

So the controller shall implement a random number generator, but there's no entropy device available yet, so it cannot do that, and thus this fake RNG is implemented? 
It's unclear to me why we aren't fixing the underlying issue first, instead of adding a fake RNG. The requirement is seemingly not satisfied with this fake implementation anyways, so it's not solving any problems anyways, which is why it's unclear to me why it's even being added. 

To me it seems like it's half-solving an issue that isn't requested, and the requirements are undocumented and yet seemingly not satisfied

cvinayak commented:
> To me it seems like it's half-solving an issue that isn't requested, and the requirements are undocumented and yet seemingly not satisfied

I believe in something is better than nothing, which is the case now where there isnt any sort of entropy in the experimental port for nRF54L SoC (bsim target use the FAKE RNG driver in Zephyr). This is not about half-solving, but incrementally reaching a goal.

> `lll_rand_isr_get` is only required to return a Pseudo-Random Number Generator (PRNG), and quickly. It doesn't need to be cryptographically secure, and it would even be disadvantageous in some cases (a CSPRNG may be slower).

There are both PRNG and CSPRNG requirement in the Controller implementation. You are free to send your suggestions as a PR, bsim test in this PR will cover it and it will be better than the quick one added in this PR.

ajf58 commented:
> I believe in something is better than nothing, which is the case now where there isnt any sort of entropy in the experimental port for nRF54L SoC (bsim target use the FAKE RNG driver in Zephyr).

If badly implemented this will, at best, create a bad citizen in the very congested 2.4GHz band.

BabbleSim is for simulation; it doesn't transmit/receive over RF. Being able to re-run simulations and get identical outputs is useful. Isn't also using an existing pseudo-random number generator?

The number generator is seeding with 0[1], so every nRF54L SoC will hop through the same sequence. The more nRF54L in the same RF environment, the more likely a collision. As soon as that occurs, the backoff delays will be in lockstep, and I strongly suspect higher layers will disconnect.

> This is not about half-solving, but incrementally reaching a goal.

Draft PR status to share visibility, but otherwise keep it in a fork? I don't think this is ready (or ever appropriate) to merge into `main` due to the concerns raised by others and me.

> You are free to send your suggestions as a PR, bsim test in this PR will cover it and it will be better than the quick one added in this PR.

I think there's enough folk at Nordic, many who are employed to work on BLE, who can do this.

I don't feel I need to solve this problem, but I can comment to suggest this is not suitable for merging. 

[1] https://github.com/zephyrproject-rtos/zephyr/pull/78434/files#diff-c07d44dc08255eb3ba8efd8480af6a8cf2ec353e2a9ee4f514470ae944ee121eR192


Draft PR status

cvinayak commented:
> The number generator is seeding with 0[1], so every nRF54L SoC will hop through the same sequence.

Provided all devices are powered on together, and all have the same clock source. Otherwise, each will drift apart. Also, the fake value will be seeded by the radio reception jitter in this PR.

> Draft PR status to share visibility, but otherwise keep it in a fork? I don't think this is ready (or ever appropriate) to merge into main due to the concerns raised by others and me.

Sure. I do not need to keep it ready to review or in draft, as I will not be able to contribute a real solution in this PR.

> I think there's enough folk at Nordic, many who are employed to work on BLE, who can do this.
I don't feel I need to solve this problem, but I can comment to suggest this is not suitable for merging. 

Not all contributions or comments are from employed community member, and I am always thankful to many features that have been contributed so far.
